File: bindings/node/examples/fluent_query.ts
import path from 'node:path'

import { and, Database, between, eq, inList, not } from '..'
import { DemoNode, isDemoNode } from './shared'

const DEFAULT_DB_PATH = './fluent-query.db'

function invokedDirectly(scriptBase: string): boolean {
  const entry = process.argv[1]
  if (!entry) {
    return false
  }
  return path.basename(entry).startsWith(scriptBase)
}

export async function runFluentQueryExample(dbPath: string = DEFAULT_DB_PATH): Promise<void> {
  const db = Database.open(dbPath, { autocheckpointMs: 0 })
  db.seedDemo()

  const scalarRows = await db
    .query()
    .nodes('User')
    .where(
      and(
        inList('name', ['Ada', 'Grace', 'Alan']),
        not(eq('name', 'Alan')),
      ),
    )
    .select('name')
    .execute()
  console.log('names returned by nodes() scope:')
  for (const row of scalarRows) {
    if (typeof row.name === 'string') {
      console.log(`- ${row.name}`)
    }
  }

  const { rows: followerRows, request_id: requestId } = await db
    .query()
    .match({ followee: 'User', follower: 'User' })
    .where('FOLLOWS', { var: 'followee', label: 'User' })
    .on('follower', (scope) => scope.where(eq('name', 'Ada')))
    .on('followee', (scope) => scope.where(between('name', 'Ada', 'Grace')))
    .select([
      { var: 'follower', as: 'follower' },
      { var: 'followee', as: 'followee' },
    ])
    .requestId('fluent-query')
    .distinct()
    .execute(true)
  console.log('follow relationships returned by match().on() scopes:')
  for (const row of followerRows) {
    const follower = row.follower
    const followee = row.followee
    if (isDemoNode(follower) && isDemoNode(followee)) {
      const followerName = String(follower.props.name ?? '(unknown)')
      const followeeName = String(followee.props.name ?? '(unknown)')
      console.log(`source=${follower._id} (${followerName}) -> target=${followee._id} (${followeeName})`)
    }
  }
  console.log(`request id for previous query: ${requestId}`)
}

if (invokedDirectly('fluent_query')) {
  runFluentQueryExample(process.argv[2] ?? DEFAULT_DB_PATH).catch((err) => {
    console.error(err)
    process.exit(1)
  })
}

File: bindings/python/examples/fluent_query.py
"""End-to-end fluent query example aligned with the canonical Stage 8 payload."""

from __future__ import annotations

import tempfile
from pathlib import Path

from sombra import Database
from sombra.query import and_, between, eq, in_list, not_


def temp_db_path() -> str:
    directory = Path(tempfile.mkdtemp())
    return str(directory / "fluent-query.db")


def main() -> None:
    db = Database.open(temp_db_path())
    db.seed_demo()

    scalar_rows = (
        db.query()
        .nodes("User")
        .where(and_(in_list("name", ["Ada", "Grace", "Alan"]), not_(eq("name", "Alan"))))
        .select("name")
        .execute()
    )
    print("names returned by nodes() scope:")
    for row in scalar_rows:
        print(f"- {row['name']}")

    payload = (
        db.query()
        .match({"followee": "User", "follower": "User"})
        .where("FOLLOWS", {"var": "followee", "label": "User"})
        .on("follower", lambda scope: scope.where(eq("name", "Ada")))
        .on("followee", lambda scope: scope.where(between("name", "Ada", "Grace")))
        .select(["follower", "followee"])
        .request_id("fluent-query")
        .distinct()
        .execute(with_meta=True)
    )

    print("follow relationships returned by match().on() scopes:")
    for row in payload.rows():
        src = row["follower"]
        dst = row["followee"]
        print(
            f"source={src['_id']} ({src['props'].get('name')}) "
            f"-> target={dst['_id']} ({dst['props'].get('name')})"
        )
    print(f"request id for previous query: {payload.request_id()}")


if __name__ == "__main__":
    main()
